src/Prefetch.cpp

```diff
@@ -0,0 +1,280 @@
+
+            // Record let assignments
+            // except for stride, already defined elsewhere
+            plets.push_back(make_pair(min_name, box[i].min));
+            plets.push_back(make_pair(max_name, box[i].max));
+        }
+
+        // Create a buffer_t object for this prefetch.
+        vector<Expr> args;
+
+        Expr first_elem = Load::make(t, varname, 0, BufferPtr(), Parameter());
+        args.push_back(Call::make(Handle(), Call::address_of, {first_elem}, Call::PureIntrinsic));
+        args.push_back(make_zero(t));
+        for (int i = 0; i < dims; i++) {
+            args.push_back(min_var[i]);
+            args.push_back(max_var[i] - min_var[i] + 1);
+            args.push_back(stride_var[i]);
+        }
+
+        // Create the create_buffer_t call
+        Expr prefetch_buf = Call::make(type_of<struct buffer_t *>(), Call::create_buffer_t,
+                              args, Call::Intrinsic);
+        plets.push_back(make_pair(varname_prefetch_buf, prefetch_buf));
+
+        // Create the prefetch call
+        Expr num_elem = stride_var[dims-1] * extent_var[dims-1];
+        vector<Expr> args_prefetch = {
+            dims,
+            elem_size_bytes,
+            num_elem,
+            var_prefetch_buf
+        };
+        Stmt stmt_prefetch = Evaluate::make(Call::make(Int(32), Call::prefetch_buffer_t,
+                              args_prefetch, Call::Intrinsic));
+        // TODO: Opt: Generate more control code for prefetch_buffer_t in Prefetch.cpp?
+        // TODO       Passing box info through a buffer_t results in ~30 additional stores/loads
+
+        pstmts.push_back(stmt_prefetch);
+
+        // No guard needed, address range checked in prefetch runtime
+        Stmt pbody = Block::make(pstmts);
+        for (size_t i = plets.size(); i > 0; i--) {
+            pbody = LetStmt::make(plets[i-1].first, plets[i-1].second, pbody);
+        }
+
+        return pbody;
+    }
+
+    void visit(const Let *op) {
+        Interval in = bounds_of_expr_in_scope(op->value, intervals);
+        intervals.push(op->name, in);
+        IRMutator::visit(op);
+        intervals.pop(op->name);
+    }
+
+    void visit(const LetStmt *op) {
+        Interval in = bounds_of_expr_in_scope(op->value, intervals);
+        intervals.push(op->name, in);
+        IRMutator::visit(op);
+        intervals.pop(op->name);
+    }
+
+    void visit(const For *op) {
+        Stmt body = op->body;
+
+        string func_name = tuple_func(op->name);
+        string loop_var_name = tuple_var(op->name);
+        const vector<Prefetch> &prefetches = get_func(func_name).schedule().prefetches();
+
+        // Add loop variable to interval scope for any inner loop prefetch
+        Expr loop_var = Variable::make(Int(32), op->name);
+        Interval prein(loop_var, loop_var);
+        intervals.push(op->name, prein);
+
+        body = mutate(body);
+
+        for (const Prefetch &p : prefetches) {
+            if (p.var != loop_var_name) {
```

Comment:
I think the right way to do this is to say:

    if (!ends_with(op->name, p.var))

Then you can probably get rid of tuple_var.